### PR TITLE
Nuclear operatives can now be other species. Pref toggle to always be human as operative

### PR DESCRIPTION
**Pull request by Rhials**
## About The Pull Request

image

When #16075 was created in the ancient year of 2016, it was a stopgap solution to plasmamen spawning in with a regular nukie outfits and burning to death. As a result, nukies have been human-only for the past seven years. I come to you today to remove the stopgap, and once again enable non-human nukies.

This was approved in the past (#59590), and I'm assuming that approval is still valid. Let's take a look at how things work:

Spawning in as a Nuclear Operative (leader included), Lone Operative, Clown Operative, or Reinforcement will now set your species to the one if your prefs. That means lizard, cat, plasma, and ethereal nukies are all on the table now.

Plasmamen now have their own syndicate envirohelm/suit combo. It doesn't look fantastic (it's just a palette swapped roboticists envirosuit) but I think it's serviceable at least. When the operative antag datum is given, and the owner is a plasmamen, they will instead be given an alternate outfit with these equipped. Loneop plasmamen will instead be given a MODsuit with the plasma stabilizer MODule installed. These MODules also spawn in the firebase lockers in case they're needed.

**But what about the AI?** Fortunately for the balancebrains in the crowd, there is a preference (enabled by default!) to always roll human as an operative. The AI still has to reasonably assume that nukies are human until proven otherwise so please don't think this will affect the rules of engagement for anyone not _directly opting into it._

image

(This pref will only appear if you have a nukie job enabled in the antag menu!)

Huge thanks to Melbert who let me peek at his code on a branch trying to do the same thing. My approach ended up being entirely on the antag datum, with nothing on the job datum.
## Why It's Good For The Game

Rectifies a curious change made long long ago. Expands customization to nukies while not affecting balance for those who care.
## Changelog
:cl: Rhials
add: Nuclear operatives have expanded the scope of their hiring practices. Members of all species can now roll nukie!
qol: The preferences menu now has an option to always be human when being selected for a nuclear operative role. Check it out!
image: Plasmaman operative envirosuits.
/:cl:

### CODE REVIEW

**Comment by Jackal-boop** [comment]
this is good

**Comment by mc-oofert** [comment]
if my leader was a nonhuman i would probably call it a guaranteed flukeop

**Comment by Livrah** [comment]
Like flukie teammates wasn't bad enough already, some races take extra dmg.

**Comment by Ticklezcat** [comment]
> Like flukie teammates wasn't bad enough already, some races take extra dmg.

Securities weakest lasers vs. The Syndicate's strongest moff.

**Comment by DaydreamIQ** [comment]
As a human-only player, this is pretty funny. Plasmeme ops might be weaker but it makes their win all the stronger.

File: code/modules/antagonists/nukeop/nukeop.dm [hunk 1]
```diff
@@ -35,9 +41,15 @@
 	if(!nukeop_outfit) // this variable is null in instances where an antagonist datum is granted via enslaving the mind (/datum/mind/proc/enslave_mind_to_creator), like in golems.
 		return
 
-	operative.set_species(/datum/species/human) //Plasmamen burn up otherwise, and besides, all other species are vulnerable to asimov AIs. Let's standardize all operatives being human.
+	// If our nuke_ops_species pref is set to TRUE, (or we have no client) make us a human
+	if(!operative.client || operative.client.prefs.read_preference(/datum/preference/toggle/nuke_ops_species))
+		operative.set_species(/datum/species/human)
+
+	if(is_species(operative, /datum/species/plasmaman))
+		operative.equipOutfit(plasmaman_outfit)
```
Comment (MrMelbert) [inline code/modules/antagonists/nukeop/nukeop.dm:49]: This should not be a hardcoded thing, it should defer to the species, IE via `pre_equip_species_outfit`. 
If we (or downstreams) wish to add a new species which requires its own unique outfits, we shouldn't have to come back to this to add `isplasmaman, elsif snowflakeman, elseif fishman, else...`

(this is the reason why I used jobs datums, it worked with that proc nicely)

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/_common/antag_spawner.dm [hunk 2]
```diff
@@ -157,7 +159,13 @@
 
 	antag_datum = new()
 	antag_datum.send_to_spawnpoint = FALSE
+
+	if(!our_client.prefs.read_preference(/datum/preference/toggle/nuke_ops_species))
+		var/species_type = our_client.prefs.read_preference(/datum/preference/choiced/species)
+		nukie.set_species(species_type)
```
Comment (MrMelbert) [inline code/modules/antagonists/_common/antag_spawner.dm:165]: This feels backwards or unnecessary. 
`safe_transfer_prefs_to` should already set them to their pref species.

**Comment by bruhlookatthisdood** [comment]
i feel like you should be able to choose what species you play as

**Comment by moocowswag** [comment]
non human nukies do have advantages though, lizard gets tackle resistance so its slightly safer to go desword / shielded modsuit.

plasmameme can try to do a singuloose/tesloose sm delam and try to avoid direct combat all together and just make the station a bad time for the crew.

Plasmameme`s downside is also irrelevant if played safely with shielded modsuit...atleast until bombs come into play.

Ethereal....cant revive crystal because nukies gib themselves on death :(

Don`t moffs see better in the dark?

**Comment by Rex9001** [comment]
> non human nukies do have advantages though, lizard gets tackle resistance so its slightly safer to go desword / shielded modsuit.
> 
> 
> 
> plasmameme can try to do a singuloose/tesloose sm delam and try to avoid direct combat all together and just make the station a bad time for the crew.
> 
> 
> 
> Plasmameme`s downside is also irrelevant if played safely with shielded modsuit...atleast until bombs come into play.
> 
> 
> 
> Ethereal....cant revive crystal because nukies gib themselves on death :(
> 
> 
> 
> Don`t moffs see better in the dark?
> 
> 
> 
> And felinids can catch the lasers for the other nukies like a cat trying to chase a laser pointer that kills them.

You can disable the microbomb implant with the atropine injector in your box

**Comment by Googles-Hands** [comment]
Mirror NT's Head of Staff and restrict nuke ops leaders to human only. :)

**Comment by san7890** [comment]
ready for review when dehardcoding is done

**Comment by necromanceranne** [comment]
Woo, plasmaman ops!

Besure to have plasmaman ops spawn with plasma stabilizers to put into their modsuits.

**Comment by SimplySmiley123** [comment]
> non human nukies do have advantages though, lizard gets tackle resistance so its slightly safer to go desword / shielded modsuit.
> 
> plasmameme can try to do a singuloose/tesloose sm delam and try to avoid direct combat all together and just make the station a bad time for the crew.
> 
> Plasmameme`s downside is also irrelevant if played safely with shielded modsuit...atleast until bombs come into play.
> 
> Ethereal....cant revive crystal because nukies gib themselves on death :(
> 
> Don`t moffs see better in the dark?
> 
> And felinids can catch the lasers for the other nukies like a cat trying to chase a laser pointer that kills them.

Honestly adds some variety with some downsides to boot.
Like ethereals and plasmamen taking more direct damage and all of the non-humans not being protected from Shitsimov.

**Comment by Fritzlschnitzel** [comment]
The syndicate has fallen.

**Comment by ExcessiveUseOfCobblestone** [comment]
> Some players play hardcore random as antagonists already, doubt taking a lizard specie is gonna do any worse then having a paraplegic and blind leader with a brain tumor and blood deficiency.

intentionally sandbagging your team seems like something that should be curbed or at least agreed upon by the people who have to deal with it. could care less for solo antags but not a fan of people taking purposefully debilitating traits on the team mode.

**Comment by Rhials** [comment]
I've pushed a solution that doesn't require any funky if checks on the outfit distribution side.

The current solution comes in two parts, the outfit_override_registry list and the equip_species_outfit proc.

The outfit_override_registry is a list, wherein the key is a default outfit path, and the value is the outfit you'd like to replace it with when equipping a species holder with the key outfit.

For instance, in the plasmaman registry the key/value pair is `/datum/outfit/syndicate/reinforcement = /datum/outfit/syndicate/reinforcement/plasmaman`, so when the reinforcement outfit is equipped via `equip_species_outfit()` proc, it will equip the `/datum/outfit/syndicate/reinforcement/plasmaman` outfit instead of the default.

This solution should be scalable to other species, and other use cases, without any issue.

Testing hasn't shown any differences in functionality -- this worked WAY better than I expected it to.

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 3]
```diff
@@ -385,6 +385,22 @@
...
+/mob/living/carbon/human/proc/equip_species_outfit(outfit, visualsOnly = FALSE)
+	var/datum/outfit/outfit_to_equip
+
+	var/override_outfit_path = dna?.species.outfit_override_registry[outfit]
+	if(override_outfit_path)
+		outfit_to_equip = new override_outfit_path
+	else
+		outfit_to_equip = new outfit
+
+	if(!outfit_to_equip)
```
Comment (san7890) [inline code/modules/mob/living/carbon/human/inventory.dm:398]: `isnull()` here

File: code/modules/mob/living/carbon/human/species_types/plasmamen.dm [hunk 4]
```diff
@@ -51,6 +51,13 @@
 
 	ass_image = 'icons/ass/assplasma.png'
 
+	outfit_override_registry = list(
+		/datum/outfit/syndicate = /datum/outfit/syndicate/plasmaman,
+		/datum/outfit/syndicate/leader = /datum/outfit/syndicate/leader/plasmaman,
+		/datum/outfit/syndicate/reinforcement = /datum/outfit/syndicate/reinforcement/plasmaman,
+		/datum/outfit/syndicate/full = /datum/outfit/syndicate/full/plasmaman,
```
Comment (san7890) [inline code/modules/mob/living/carbon/human/species_types/plasmamen.dm:58]: alphabetize for better extensibility?

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/nukeop/nukeop.dm [hunk 5]
```diff
@@ -35,9 +39,12 @@
 	if(!nukeop_outfit) // this variable is null in instances where an antagonist datum is granted via enslaving the mind (/datum/mind/proc/enslave_mind_to_creator), like in golems.
 		return
 
-	operative.set_species(/datum/species/human) //Plasmamen burn up otherwise, and besides, all other species are vulnerable to asimov AIs. Let's standardize all operatives being human.
+	// If our nuke_ops_species pref is set to TRUE, (or we have no client) make us a human
+	if(!operative.client || operative.client.prefs.read_preference(/datum/preference/toggle/nuke_ops_species))
```
Comment (san7890) [inline code/modules/antagonists/nukeop/nukeop.dm:43]:
```suggestion
	if(isnull(operative.client) || operative.client.prefs.read_preference(/datum/preference/toggle/nuke_ops_species))
```

didn't see this my bad

**Comment by Livrah** [comment]
Can non human species get a macro bomb implant instead of regular one?

**Pull request merged**